### PR TITLE
Support advanced indexing

### PR DESCRIPTION
**Pull request by kaushikcfd**
- Adds support for non-1 step slices
- Adds slices for indexing via integer arrays into pt arrays

See https://numpy.org/doc/stable/reference/arrays.indexing.html#advanced-indexing for the partial spec.

### CODE REVIEW

File: pytato/utils.py [hunk 1]
```diff
@@ -258,3 +261,277 @@ def are_shapes_equal(shape1: ShapeType, shape2: ShapeType) -> bool:
...
+    """
+    Returns *True* iff it can be proven that ``expr >= 0``.
+    """
+    if isinstance(expr, int):
+        return expr >= 0
+
+    assert isinstance(expr, Array)
+    from pytato.transform import InputGatherer
+    # type-ignore reason: passed Set[Optional[str]]; function expects Set[str]
+    space = _create_size_param_space({expr.name  # type: ignore
```
Comment (inducer) [inline pytato/utils.py:320]: All `type: ignore`s should be `type: ignore[code]`.

File: pytato/utils.py [hunk 2]
```diff
@@ -258,3 +261,277 @@ def are_shapes_equal(shape1: ShapeType, shape2: ShapeType) -> bool:
...
+
+
+def _is_positive(expr: ShapeComponent) -> bool:
+    """
+    Returns *True* iff it can be proven that ``expr >= 0``.
+    """
+    if isinstance(expr, int):
+        return expr >= 0
+
+    assert isinstance(expr, Array)
```
Comment (inducer) [inline pytato/utils.py:317]: Should probably also check the shape (i.e. scalar-ness) of the array.

File: pytato/utils.py [hunk 3]
```diff
@@ -258,3 +261,277 @@ def are_shapes_equal(shape1: ShapeType, shape2: ShapeType) -> bool:
...
+
+def _get_size_params_assumptions_bset(space: isl.Space) -> isl.BasicSet:
+    bset = isl.BasicSet.universe(space)
+    for name in bset.get_var_dict():
+        bset = bset.add_constraint(isl.Constraint.ineq_from_names(space, {name: 1}))
+
+    return bset
+
+
+def _is_positive(expr: ShapeComponent) -> bool:
```
Comment (inducer) [inline pytato/utils.py:310]:
```suggestion
def _is_nonnegative(expr: ShapeComponent) -> bool:
```

File: pytato/utils.py [hunk 4]
```diff
@@ -258,3 +261,277 @@ def are_shapes_equal(shape1: ShapeType, shape2: ShapeType) -> bool:
...
+    # type-ignore reason: passed Set[Optional[str]]; function expects Set[str]
+    space = _create_size_param_space({expr.name  # type: ignore
+                                      for expr in InputGatherer()(expr)})
+    aff = ShapeToISLExpressionMapper(space)(expr)
+    # type-ignore reason: mypy doesn't know comparing isl.Sets returns bool
+    return (aff.ge_set(aff * 0)  # type: ignore
+            <= _get_size_params_assumptions_bset(space))
+
+
+def _is_negative(expr: ShapeComponent) -> bool:
```
Comment (inducer) [inline pytato/utils.py:328]:
```suggestion
def _is_nonpositive(expr: ShapeComponent) -> bool:
```

File: pytato/utils.py [hunk 5]
```diff
@@ -258,3 +261,277 @@ def are_shapes_equal(shape1: ShapeType, shape2: ShapeType) -> bool:
...
+    """
+    Returns *True* iff it can be proven that ``expr <= 0``.
+    """
+    if isinstance(expr, int):
```
Comment (inducer) [inline pytato/utils.py:332]: Why not implement this by `_is_nonnegative` on the input with a flipped sign?

File: pytato/utils.py [hunk 6]
```diff
@@ -258,3 +261,277 @@ def are_shapes_equal(shape1: ShapeType, shape2: ShapeType) -> bool:
...
+Tfind = TypeVar("Tfind")
+
+
+def _find(seq: Sequence[Tfind], key: Callable[[Tfind], bool]) -> int:
+    i = 0
+    for val in seq:
+        if key(val):
+            break
+        i += 1
+    return i
```
Comment (inducer) [inline pytato/utils.py:357]: * Use `enumerate`?
* Rename `key` -> `predicate`?
* `raise` if not found?

File: test/test_codegen.py [hunk 7]
```diff
@@ -1034,6 +1034,34 @@ def test_vdot(ctx_factory, a_shape, b_shape, a_dtype, b_dtype):
...
 
 
+def test_advanced_indexing(ctx_factory):
+    from numpy.random import default_rng
+
+    ctx = ctx_factory()
+    cq = cl.CommandQueue(ctx)
+
+    rng = default_rng()
+    x_in = rng.random((3, 3, 3, 3))
```
Comment (inducer) [inline test/test_codegen.py:1044]: IMO, 3 is not nearly big enough to cover all the corner cases for slices.

**Review by inducer — COMMENTED** [review]
Thanks for working on this! Here are some initial comments.

Note to future self: I've read everything but `utils` below `_normalize_slice`.

**Comment by inducer** [comment]
Unsubscribing... @-mention or request review once it's ready for a look or needs attention.

File: pytato/array.py [hunk 8]
```diff
@@ -399,85 +399,17 @@ def __len__(self) -> ShapeComponent:
 
     def __getitem__(self,
             slice_spec: Union[SliceItem, Tuple[SliceItem, ...]]) -> Array:
+        """
+        .. warning::
+
+            Out-of-bound accesses via :class:`Array` indices is undefined
+            behavior and may pass silently.
```
[the remark this answers is not in the document]
Comment (kaushikcfd) [reply pytato/array.py]: Done, thanks!

File: pytato/utils.py (hunk 1 above)
Comment (kaushikcfd) [reply pytato/utils.py:320]: Done, thanks!

File: pytato/utils.py (hunk 3 above)
Comment (kaushikcfd) [reply pytato/utils.py]: Agreed that's a better name. Fixed.

File: pytato/utils.py (hunk 4 above)
Comment (kaushikcfd) [reply pytato/utils.py]: Done, thanks!

File: pytato/utils.py (hunk 5 above)
Comment (kaushikcfd) [reply pytato/utils.py]: Oops :disappointed: (fixed!)

File: pytato/utils.py (hunk 6 above)
Comment (kaushikcfd) [reply pytato/utils.py:346]: Its downstream user needs it this way. I.e. not raise but give the index and give the length of the array when not found.

File: pytato/utils.py (hunk 2 above)
Comment (kaushikcfd) [reply pytato/utils.py]: Yep, thx!

File: pytato/utils.py [hunk 9]
```diff
@@ -258,3 +261,277 @@ def are_shapes_equal(shape1: ShapeType, shape2: ShapeType) -> bool:
...
+    # (c) x[:, idx1, idx2, :] has a shape of (3, 2, 2, 3)
+    # (d) x[:, idx1, :, idx2] has a shape of (2, 2, 3, 3)
+    # Case (d)'s shape seems completely arbitrary => do not try to reverse
+    # engineer
+
+    for i in range(indirection_start, indirection_start+len(arys)):
+        if not isinstance(index[i], Array):
+            raise NotImplementedError("Non-contiguous sequences of array "
+                                      "indices is not properly defined in numpy"
+                                      "-spec.")
```
[the remark this answers is not in the document]
Comment (kaushikcfd) [reply pytato/utils.py]: Went ahead and implemented it anyway. Not disliking it that much now. (Stockholm syndrome?)

File: pytato/utils.py (hunk 6 above)
Comment (kaushikcfd) [reply pytato/utils.py]: This helper is no longer needed. Deleted it. (Resolving this issue).

File: test/test_codegen.py (hunk 7 above)
Comment (kaushikcfd) [reply test/test_codegen.py:1065]: Agreed, thanks. Added a fuzz tester in 78508e3. That did help me fix a corner case.

File: test/test_codegen.py [hunk 10]
```diff
@@ -1034,6 +1034,34 @@ def test_vdot(ctx_factory, a_shape, b_shape, a_dtype, b_dtype):
...
+    evt, (pt_out,) = pt.generate_loopy(x[:, ::-1, idx1, idx2])(cq)
+    np.testing.assert_allclose(pt_out, x_in[:, ::-1, idx1_in, idx2_in])
+
+    # Case 2
+    evt, (pt_out,) = pt.generate_loopy(x[-4:4:-1, idx1, idx2, :])(cq)
+    np.testing.assert_allclose(pt_out, x_in[-4:4:-1, idx1_in, idx2_in, :])
+
+    # Case 3
+    evt, (pt_out,) = pt.generate_loopy(x[idx1, idx2, -2::-1, :])(cq)
+    np.testing.assert_allclose(pt_out, x_in[idx1_in, idx2_in, -2::-1, :])
```
[the remark this answers is not in the document]
Comment (kaushikcfd) [reply test/test_codegen.py:1083]: Kept these tests. On top of it also added a fuzz tester: 78508e3.

**Comment by kaushikcfd** [comment]
@inducer: This is ready! This was slightly shorter earlier because we used to very greedily generate IndexLambdas for the indexing expressions. But that turned out to be a bad decision especially if we wanted to do some downstream analysis on the expression graphs. Took the approach of giving the index expressions a type of their own and translating them to index lambdas during code generation.

File: pytato/array.py [hunk 11]
```diff
@@ -1404,31 +1352,119 @@ def shape(self) -> Tuple[int, ...]:
...
+        assert any(isinstance(idx, Array) for idx in self.indices)
+        from pytato.utils import (get_shape_after_broadcasting,
+                                  _normalized_slice_len)
+
+        i_adv_indices = tuple(i
+                              for i, idx_expr in enumerate(self.indices)
+                              if isinstance(idx_expr, (Array, int)))
+        i_basic_indices = tuple(i
+                                for i, idx_expr in enumerate(self.indices)
+                                if isinstance(idx_expr, NormalizedSlice))
```
[the remark this answers is not in the document]
Comment (kaushikcfd) [reply pytato/array.py:1415]: Added `pytato.utils.partition`.  We would have to introduce dependency on [more_itertools](https://pypi.org/project/more-itertools/) otherwise.

File: pytato/array.py [hunk 12]
```diff
@@ -1404,31 +1352,119 @@ def shape(self) -> Tuple[int, ...]:
...
+    @property
+    def shape(self) -> ShapeType:
+        assert len(self.indices) == self.array.ndim
+        assert any(isinstance(idx, Array) for idx in self.indices)
+        from pytato.utils import (get_shape_after_broadcasting,
+                                  _normalized_slice_len)
+
+        i_adv_indices = tuple(i
+                              for i, idx_expr in enumerate(self.indices)
+                              if isinstance(idx_expr, (Array, int)))
```
[the remark this answers is not in the document]
Comment (kaushikcfd) [reply pytato/array.py:1412]: When an indexing expression has both slices/ellipsis Numpy only considers slices as basic indices.

From https://numpy.org/doc/stable/reference/arrays.indexing.html#combining-advanced-and-basic-indexing
> There are two parts to the indexing operation, the subspace defined by the basic indexing (excluding integers) and the subspace from the advanced indexing part.

Also, the example below confirms it.
```python
In [2]: a = np.random.rand(3, 3, 3, 3)

In [3]: a[:, [0, 2], :, 1].shape
Out[3]: (2, 3, 3)  # '1' was treated as an advanced index.

In [4]: a[:, [0, 2], :, :].shape
Out[4]: (3, 2, 3, 3)
```

File: pytato/array.py [hunk 13]
```diff
@@ -282,6 +286,13 @@ def _truediv_result_type(arg1: DtypeOrScalar, arg2: DtypeOrScalar) -> np.dtype[A
         return dtype
 
 
+@dataclass(frozen=True, eq=True)
+class NormalizedSlice:
```
[the remark this answers is not in the document]
Comment (kaushikcfd) [reply pytato/array.py:290]: Done.

File: pytato/array.py [hunk 14]
```diff
@@ -1404,31 +1352,119 @@ def shape(self) -> Tuple[int, ...]:
...
+                                for i, idx_expr in enumerate(self.indices)
+                                if isinstance(idx_expr, NormalizedSlice))
+
+        assert not any(i_adv_indices[0] < i_basic_idx < i_adv_indices[-1]
+                       for i_basic_idx in i_basic_indices)
+
+        adv_idx_shape = get_shape_after_broadcasting([self.indices[i_idx]
+                                                      for i_idx in i_adv_indices])
+
+        # type-ignored because mypy cannot figure out basic-indices only refer slices
```
[the remark this answers is not in the document]
Comment (kaushikcfd) [reply pytato/array.py:1423]: We would need to add `typing.cast` to self.indices[i_idx]`, but was turning out to be noisy.

File: pytato/array.py [hunk 15]
```diff
@@ -1404,31 +1352,119 @@ def shape(self) -> Tuple[int, ...]:
...
+        pre_basic_idx_shape = tuple(_normalized_slice_len(self.indices[i_idx])  # type: ignore[arg-type]  # noqa: E501
+                                    for i_idx in i_basic_indices
+                                    if i_idx < i_adv_indices[0])
+
+        # type-ignored because mypy cannot figure out basic-indices only refer slices
```
[the remark this answers is not in the document]
Comment (kaushikcfd) [reply pytato/array.py:1428]: Thanks, done!

File: test/test_codegen.py [hunk 16]
```diff
@@ -1055,6 +1055,97 @@ def test_reduction_adds_deps(ctx_factory):
...
+    evt, (pt_out,) = pt.generate_loopy(x[:, idx1, -2::-1, idx2])(cq)
+    np.testing.assert_allclose(pt_out, x_in[:, idx1_in, -2::-1, idx2_in])
+
+
+def test_advanced_indexing_fuzz(ctx_factory):
+    from numpy.random import default_rng
+
+    ctx = ctx_factory()
+    cq = cl.CommandQueue(ctx)
+    rng = default_rng(0)
```
[the remark this answers is not in the document]
Comment (kaushikcfd) [reply test/test_codegen.py:1095]: Thanks!

File: pytato/codegen.py [hunk 17]
```diff
@@ -411,11 +410,142 @@ def _indices_for_reshape(self, expr: Reshape) -> SymbolicIndex:
...
+                if isinstance(axis_len, int):
+                    bnd_name = vng("in")
+                    bindings[bnd_name] = self.rec(idx)
+                    indices.append(prim.Subscript(
+                                        prim.Variable(bnd_name),
+                                        get_indexing_expression(idx.shape,
+                                                                adv_idx_shape))
+                                   % axis_len)
+                else:
+                    raise NotImplementedError
```
[the remark this answers is not in the document]
Comment (kaushikcfd) [reply pytato/codegen.py:540]: Thanks, done.

File: pytato/utils.py [hunk 18]
```diff
@@ -258,3 +264,251 @@ def are_shapes_equal(shape1: ShapeType, shape2: ShapeType) -> bool:
...
+        if _is_non_negative(start - stop):
+            if _is_divisible(start - stop, -step):
+                return (start - stop) // (-step)
+            else:
+                return (start - stop) // (-step) + 1
+        elif _is_non_positive(start - stop):
+            return 0
+        else:
+            # ISL could not ascertain the expression's sign
+            raise NotImplementedError
```
[the remark this answers is not in the document]
Comment (kaushikcfd) [reply pytato/utils.py:419]: Done.

File: pytato/utils.py [hunk 19]
```diff
@@ -258,3 +264,251 @@ def are_shapes_equal(shape1: ShapeType, shape2: ShapeType) -> bool:
...
+
+def _normalized_slice_len(slice_: NormalizedSlice) -> ShapeComponent:
+    start, stop, step = slice_.start, slice_.stop, slice_.step
+
+    if step > 0:
+        if _is_non_negative(stop - start):
+            if _is_divisible(stop - start, step):
+                return (stop - start) // step
+            else:
+                return (stop - start) // step + 1
```
[the remark this answers is not in the document]
Comment (kaushikcfd) [reply pytato/utils.py:403]: Agreed,much better. Done, thanks!

File: pytato/codegen.py [hunk 20]
```diff
@@ -411,11 +410,142 @@ def _indices_for_reshape(self, expr: Reshape) -> SymbolicIndex:
...
+                    bnd_name = vng("in")
+                    bindings[bnd_name] = self.rec(idx)
+                    indices.append(prim.Subscript(
+                                        prim.Variable(bnd_name),
+                                        get_indexing_expression(
+                                                idx.shape,
+                                                (1,)*i_adv_indices[0]+adv_idx_shape))
+                                   % axis_len)
+                else:
+                    raise NotImplementedError
```
[the remark this answers is not in the document]
Comment (kaushikcfd) [reply pytato/codegen.py]: Thanks, done.

File: test/test_codegen.py [hunk 21]
```diff
@@ -1055,6 +1055,97 @@ def test_reduction_adds_deps(ctx_factory):
...
+    evt, (pt_out,) = pt.generate_loopy(x[-4:4:-1, idx1, idx2, :])(cq)
+    np.testing.assert_allclose(pt_out, x_in[-4:4:-1, idx1_in, idx2_in, :])
+
+    # Case 3
+    evt, (pt_out,) = pt.generate_loopy(x[idx1, idx2, -2::-1, :])(cq)
+    np.testing.assert_allclose(pt_out, x_in[idx1_in, idx2_in, -2::-1, :])
+
+    # Case 4 (non-contiguous advanced indices)
+    evt, (pt_out,) = pt.generate_loopy(x[:, idx1, -2::-1, idx2])(cq)
+    np.testing.assert_allclose(pt_out, x_in[:, idx1_in, -2::-1, idx2_in])
```
[the remark this answers is not in the document]
Comment (kaushikcfd) [reply test/test_codegen.py:1087]: Added a case (5) here.

**Comment by kaushikcfd** [comment]
@inducer: Soft ping.

File: pytato/array.py [hunk 22]
```diff
@@ -282,6 +286,28 @@ def _truediv_result_type(arg1: DtypeOrScalar, arg2: DtypeOrScalar) -> np.dtype[A
         return dtype
 
 
+@dataclass(frozen=True, eq=True)
+class NormalizedSlice:
+    """
+    A normalized version of :class:`slice`.
```
Comment (inducer) [inline pytato/array.py:293]:
```suggestion
    A normalized version of :class:`slice`. "Normalized" as explained in :attr:`start`
    and :attr:`stop`.
```

File: pytato/array.py [hunk 23]
```diff
@@ -282,6 +286,28 @@ def _truediv_result_type(arg1: DtypeOrScalar, arg2: DtypeOrScalar) -> np.dtype[A
...
+        An instance of :class:`ShapeComponent`. Normalized to satisfy the
+        relation ``0 <= start < axis_len``, where ``axis_len`` is the length of the
+        axis being sliced.
+
+    .. attribute:: stop
+
+        An instance of :class:`ShapeComponent`. Normalized to satisfy the
+        relation ``0 <= stop < axis_len``, where ``axis_len`` is the length of
+        the axis being sliced.
+    """
```
Comment (inducer) [inline pytato/array.py:306]:
```suggestion
    
    .. attribute:: step
    """
```

File: pytato/array.py (hunk 12 above)
[the remark this answers is not in the document]
Comment (inducer) [reply pytato/array.py]: Wow, TIL. Feels like a weird inconsistency (any integer index is trivially representable as a slice!), but I guess we're bound by what the boss says. :)

[the remark this answers is not in the document]
Comment (kaushikcfd) [reply pytato/array.py]: > representable as a slice

`int` index goes a step further and squeezes the axis

```python
In [1]: import numpy as np

In [2]: a = np.arange(10)

In [3]: a[7:8]
Out[3]: array([7])

In [4]: a[7]
Out[4]: 7
```

File: pytato/utils.py [hunk 24]
```diff
@@ -258,3 +286,236 @@ def are_shapes_equal(shape1: ShapeType, shape2: ShapeType) -> bool:
...
+
+class ShapeToISLExpressionMapper(Mapper):
+    """
+    Mapper that takes a shape component and returns it as :class:`isl.Aff`.
+    """
+    def __init__(self, space: isl.Space):
+        self.cache: Dict[Array, isl.Aff] = {}
+        self.space = space
+
+    def rec(self, expr: Array) -> isl.Aff:  # type: ignore
```
Comment (inducer) [inline pytato/utils.py:301]: Why type-ignored? Also, needs a mypy error code.

File: pytato/utils.py [hunk 25]
```diff
@@ -258,3 +286,236 @@ def are_shapes_equal(shape1: ShapeType, shape2: ShapeType) -> bool:
...
+    start, stop, step = slice_.start, slice_.stop, slice_.step
+    if step is None:
+        step = 1
+    if not isinstance(step, int):
+        raise ValueError(f"slice step must be an int or 'None' (got a {type(step)})")
+    if step == 0:
+        raise ValueError("slice step cannot be zero")
+
+    if step > 0:
+        _DEFAULT_START_: ShapeComponent = 0  # noqa: N806
```
Comment (inducer) [inline pytato/utils.py:375]: Why are these guys named with capitals and underscores?

File: pytato/utils.py [hunk 26]
```diff
@@ -258,3 +286,236 @@ def are_shapes_equal(shape1: ShapeType, shape2: ShapeType) -> bool:
...
+        _DEFAULT_STOP_: ShapeComponent = axis_len  # noqa: N806
+    else:
+        _DEFAULT_START_ = axis_len - 1  # noqa: N806
+        _DEFAULT_STOP_ = -1  # noqa: N806
+
+    if start is None:
+        start = _DEFAULT_START_
+    else:
+        if isinstance(axis_len, int):
+            if -axis_len <= start < axis_len:
```
Comment (inducer) [inline pytato/utils.py:385]: Only do this in the negative case?

File: pytato/utils.py [hunk 27]
```diff
@@ -258,3 +286,236 @@ def are_shapes_equal(shape1: ShapeType, shape2: ShapeType) -> bool:
...
+                start = start % axis_len
+            else:
+                start = 0
```
Comment (inducer) [inline pytato/utils.py:388]: Is this correct? IIUC, `a[20:]` would fall into this for a five-long array, and potentially start indexing at the start.

Also, I'm worried that the tests didn't catch this. Could you improve them so that they do?

File: pytato/utils.py [hunk 28]
```diff
@@ -258,3 +286,236 @@ def are_shapes_equal(shape1: ShapeType, shape2: ShapeType) -> bool:
...
+            raise NotImplementedError
+
+    if stop is None:
+        stop = _DEFAULT_STOP_
+    else:
+        if isinstance(axis_len, int):
+            if -axis_len <= stop < axis_len:
+                stop = stop % axis_len
+            else:
+                stop = axis_len
```
Comment (inducer) [inline pytato/utils.py:399]: Similar concern as above: `a[:-300]` for a five-long should have the beginning as its stopping point.

Also, I'm worried that the tests didn't catch this. Could you improve them so that they do?

**Review by inducer — COMMENTED** [review]
Thanks! Spotted a few more things, but that should be it. Good to go after.

File: pytato/array.py [hunk 29]
```diff
@@ -1431,31 +1394,117 @@ def shape(self) -> Tuple[int, ...]:
...
+                     for idx, axis_len in zip(self.indices, self.array.shape)
+                     if isinstance(idx, NormalizedSlice))
+
+
+class AdvancedIndexInContiguousAxes(IndexBase):
+    """
+    An indexing expression with at least one :class:`Array` index and all the
+    advanced indices (i.e. scalars/array) appearing contiguously in
+    :attr:`IndexBase.indices`.
+    """
```
[the remark this answers is not in the document]
Comment (kaushikcfd) [reply pytato/array.py:1449]: Done, thanks!

File: pytato/array.py [hunk 30]
```diff
@@ -1431,31 +1394,117 @@ def shape(self) -> Tuple[int, ...]:
...
+                                     if i_idx > i_adv_indices[-1])
+
+        return pre_basic_idx_shape + adv_idx_shape + post_basic_idx_shape
+
+
+class AdvancedIndexInNoncontiguousAxes(IndexBase):
+    """
+    An indexing expression with advanced indices (i.e. scalars/arrays)
+    appearing non-contiguously in :attr:`IndexBase.indices`.
+    """
```
[the remark this answers is not in the document]
Comment (kaushikcfd) [reply pytato/array.py:1493]: Done, thanks.

File: pytato/transform.py [hunk 31]
```diff
@@ -255,9 +269,22 @@ def map_roll(self, expr: Roll) -> CombineT:
     def map_axis_permutation(self, expr: AxisPermutation) -> CombineT:
         return self.combine(self.rec(expr.array))
 
-    def map_slice(self, expr: Slice) -> CombineT:
+    def _map_index_base(self, expr: IndexBase) -> CombineT:
         return self.combine(self.rec(expr.array))
```
[the remark this answers is not in the document]
Comment (kaushikcfd) [reply pytato/transform.py]: Thanks, done.

File: pytato/transform.py [hunk 32]
```diff
@@ -475,9 +515,21 @@ def _map_index_remapping_base(self, expr: IndexRemappingBase) -> None:
 
     map_roll = _map_index_remapping_base
     map_axis_permutation = _map_index_remapping_base
-    map_slice = _map_index_remapping_base
     map_reshape = _map_index_remapping_base
 
+    def map_basic_index(self, expr: BasicIndex) -> None:
+        return self._map_index_remapping_base(expr)
```
[the remark this answers is not in the document]
Comment (kaushikcfd) [reply pytato/transform.py]: Thanks, done.

File: pytato/transform.py [hunk 33]
```diff
@@ -324,8 +351,21 @@ def map_roll(self, expr: Roll) -> R:
...
+
+    def map_contiguous_advanced_index(self,
+                                      expr: AdvancedIndexInContiguousAxes
+                                      ) -> R:
+        return self._map_index_base(expr)
+
+    def map_non_contiguous_advanced_index(self,
+                                          expr: AdvancedIndexInNoncontiguousAxes
+                                          ) -> R:
+        return self._map_index_base(expr)
```
[the remark this answers is not in the document]
Comment (kaushikcfd) [reply pytato/transform.py]: Yep, was unnecessary.

File: pytato/utils.py (hunk 24 above)
Comment (kaushikcfd) [reply pytato/utils.py]: Done.

File: pytato/utils.py (hunk 25 above)
Comment (kaushikcfd) [reply pytato/utils.py]: Got rid of them

File: pytato/utils.py (hunk 26 above)
Comment (kaushikcfd) [reply pytato/utils.py:386]: It's equivalent, right? In the +ve case the modulo will be identity mapping.

File: pytato/utils.py (hunk 27 above)
Comment (kaushikcfd) [reply pytato/utils.py]: Done, and also widened the range of 'start', 'stop', 'step' of a slice to be [-2*axis_len, 2*axis_len), earlier it was [-axis_len, axis_len).

File: pytato/utils.py (hunk 28 above)
Comment (kaushikcfd) [reply pytato/utils.py]: Done, and also widened the range of 'start', 'stop', 'step' of a slice to be [-2*axis_len, 2*axis_len), earlier it was [-axIs_len, axis_len).

File: pytato/array.py (hunk 22 above)
Comment (kaushikcfd) [reply pytato/array.py]: Thanks, done!

File: pytato/array.py (hunk 23 above)
Comment (kaushikcfd) [reply pytato/array.py:309]: Done.

File: pytato/utils.py (hunk 24 above)
Comment (kaushikcfd) [reply pytato/utils.py]: left a comment, also narrowed the scope of type-ignore via the mypy code.

**Comment by inducer** [comment]
LGTM, thanks for working on this!
